### PR TITLE
Endpoint order

### PR DESCRIPTION
**Pull request by chunyu3**
# Description
Fix https://github.com/Azure/autorest.csharp/issues/3502

- order the parameter, and put endpoint the first parameter.

# Checklist

To ensure a quick review and merge, please ensure:
- [ ] The PR has a understandable title and description explaining the _why_ and _what_.
- [ ] The PR is opened in draft if not ready for review yet.
   - If opened in draft, please allocate sufficient time (24 hours) after moving out of draft for review
- [ ] The branch is recent enough to not have merge conflicts upon creation.

# Ready to Land?
- [ ] Build is completely green
   - Submissions with test failures require tracking issue and approval of a CODEOWNER
- [ ] At least one +1 review by a CODEOWNER
- [ ] All -1 reviews are confirmed resolved by the reviewer 
   - Override/Marking reviews stale must be discussed with CODEOWNERS first

### CODE REVIEW

File: src/AutoRest.CSharp/LowLevel/Output/LowLevelClient.cs [hunk 1]
```diff
@@ -156,13 +156,13 @@ private IEnumerable<ConstructorSignature> BuildPrimaryConstructors(IReadOnlyList
...
             {
-                yield return CreatePrimaryConstructor(requiredParameters.Concat(optionalToRequired).ToArray());
+                yield return CreatePrimaryConstructor(requiredParameters.Concat(optionalToRequired).OrderBy(parameter => parameter.Name != "endpoint").ToArray());
             }
             else
             {
                 foreach (var credentialField in Fields.CredentialFields)
                 {
-                    yield return CreatePrimaryConstructor(requiredParameters.Append(CreateCredentialParameter(credentialField!.Type)).Concat(optionalToRequired).ToArray());
+                    yield return CreatePrimaryConstructor(requiredParameters.Append(CreateCredentialParameter(credentialField!.Type)).Concat(optionalToRequired).OrderBy(parameter => parameter.Name != "endpoint").ToArray());
```
Comment (ArcturusZhang) [inline src/AutoRest.CSharp/LowLevel/Output/LowLevelClient.cs:165]: this reads so much like a "patch" to our code.
Could we come up with a solution that does not look like a patch?

I heard the root cause of this is that in swagger, the endpoint is always optional. Is it possible we could make it required in the `CodeModelConverter` or anywhere else and then we do not need to do anything here and fix the issue?

Comment (chunyu3) [reply src/AutoRest.CSharp/LowLevel/Output/LowLevelClient.cs:165]: hello @ArcturusZhang The root cause is the endpoint is optional or it has default value. We also need to honor the default value and use it to call the secondary constructor in primary constructor. In that way, the endpoint parameter should maintain optional with default value. We cannot just convert it as required, and ignore the default value.

File: src/AutoRest.CSharp/LowLevel/Output/LowLevelClient.cs [hunk 2]
```diff
@@ -174,32 +180,45 @@ private IEnumerable<ConstructorSignature> BuildSecondaryConstructors(IReadOnlyLi
...
+                var allarguments = arguments.Concat(optionalParametersArguments);
+                yield return CreateSecondaryConstructor(requiredParameters, allarguments.ToArray());
             }
             else
             {
                 foreach (var credentialField in Fields.CredentialFields)
                 {
-                    yield return CreateSecondaryConstructor(requiredParameters.Append(CreateCredentialParameter(credentialField!.Type)).ToArray(), optionalParametersArguments);
+                    var credentialParameter = CreateCredentialParameter(credentialField!.Type);
+                    var allarguments = arguments.Concat(new List<FormattableString>() { $"{credentialParameter.Name}" }).Concat(optionalParametersArguments);
```
Comment (lirenhe) [inline src/AutoRest.CSharp/LowLevel/Output/LowLevelClient.cs:211]: Should it be 'allArguments'?

File: src/AutoRest.CSharp/LowLevel/Output/LowLevelClient.cs [hunk 3]
```diff
@@ -156,13 +156,19 @@ private IEnumerable<ConstructorSignature> BuildPrimaryConstructors(IReadOnlyList
 
             if (Fields.CredentialFields.Count == 0)
             {
-                yield return CreatePrimaryConstructor(requiredParameters.Concat(optionalToRequired).ToArray());
+                /* order the constructor paramters as (endpoint, requiredParameters, OptionalParamters).
+                 * use the OrderBy to put endpoint as the first parameter.
+                 * */
+                yield return CreatePrimaryConstructor(requiredParameters.Concat(optionalToRequired).OrderBy(parameter => parameter.Name != "endpoint").ToArray());
```
Comment (lirenhe) [inline src/AutoRest.CSharp/LowLevel/Output/LowLevelClient.cs:162]: Do we need to care about the case when comparing 'endpoint'?

File: src/AutoRest.CSharp/LowLevel/Output/LowLevelClient.cs (hunk 2 above)
Comment (chunyu3) [reply src/AutoRest.CSharp/LowLevel/Output/LowLevelClient.cs]: resolved.

File: src/AutoRest.CSharp/LowLevel/Output/LowLevelClient.cs (hunk 3 above)
Comment (chunyu3) [reply src/AutoRest.CSharp/LowLevel/Output/LowLevelClient.cs]: `endpoint` is current hardcoded. but it is better to compare case-insensitive. updated. Thanks,

**Review by ArcturusZhang — APPROVED** [review]

File: src/AutoRest.CSharp/LowLevel/Output/LowLevelClient.cs [hunk 4]
```diff
@@ -156,13 +156,19 @@ private IEnumerable<ConstructorSignature> BuildPrimaryConstructors(IReadOnlyList
 
             if (Fields.CredentialFields.Count == 0)
             {
-                yield return CreatePrimaryConstructor(requiredParameters.Concat(optionalToRequired).ToArray());
+                /* order the constructor paramters as (endpoint, requiredParameters, OptionalParamters).
+                 * use the OrderBy to put endpoint as the first parameter.
+                 * */
+                yield return CreatePrimaryConstructor(requiredParameters.Concat(optionalToRequired).OrderBy(parameter => !parameter.Name.Equals("endpoint", StringComparison.OrdinalIgnoreCase)).ToArray());
```
Comment (pshao25) [inline src/AutoRest.CSharp/LowLevel/Output/LowLevelClient.cs:162]: We have a lot of places directly use "endpoint". I would suggest make it a constant variable.

**Review by pshao25 — APPROVED** [review]

File: src/AutoRest.CSharp/LowLevel/Output/LowLevelClient.cs (hunk 4 above)
Comment (ArcturusZhang) [reply src/AutoRest.CSharp/LowLevel/Output/LowLevelClient.cs:162]: actually we could reference the `KnownParameters` where we have a parameter created for endpoint

Comment (chunyu3) [reply src/AutoRest.CSharp/LowLevel/Output/LowLevelClient.cs:162]: Yes. Maybe we need to file a MQ task to update `endpoint` to `KnowParameters.Endpoint.Name`. And after migrate mgp to InputModel, we will update this together.

**Review by lirenhe — APPROVED** [review]

**Pull request merged**